jsapp/js/bem.ts

```diff
@@ -1,382 +1,103 @@
-bem.ProjectDownloads__title = bem.ProjectDownloads.__('title', 'span');
-bem.ProjectDownloads__textButton = bem.ProjectDownloads.__('text-button', 'button');
-bem.ProjectDownloads__selectorRow = bem.ProjectDownloads.__('selector-row');
-bem.ProjectDownloads__anonymousRow = bem.ProjectDownloads.__('anonymous-row');
-bem.ProjectDownloads__legacyIframeWrapper = bem.ProjectDownloads.__('legacy-iframe-wrapper');
-bem.ProjectDownloads__submitRow = bem.ProjectDownloads.__('submit-row', 'footer');
-bem.ProjectDownloads__exportsSelector = bem.ProjectDownloads.__('exports-selector');
-bem.ProjectDownloads__deleteSettingsButton = bem.ProjectDownloads.__('delete-settings-button', 'button');
-bem.ProjectDownloads__exportsCreator = bem.ProjectDownloads.__('exports-creator');
+interface BemComponentProps extends React.ComponentProps<any> {
+  m?: string | string[] | BemModifiersObject
+}
 
-bem.BackgroundAudioPlayer = bem.create('background-audio-player');
-bem.BackgroundAudioPlayer__label = bem.BackgroundAudioPlayer.__('label', '<label>');
-bem.BackgroundAudioPlayer__audio = bem.BackgroundAudioPlayer.__('audio', '<audio>');
+interface BemInstance extends React.ComponentClass<BemComponentProps, {}> {
+  blockName: string
+}
 
-export default bem;
+/**
+ * Creates a BEM class for block or element.
+ * For first parameter pass `null` to create a Block component,
+ * or pass existing Block component to create Element component (a child).
+ * Please no angle brackets for `htmlTagName`.
+ */
+export function makeBem(
+  parent: BemInstance | null,
+  name: string,
+  htmlTagName: string = 'div'
+): BemInstance {
+  class BemComponent extends React.Component<BemComponentProps, {}> {
+    static blockName: string = parent ? parent.blockName : name
+
+    constructor(props: BemComponentProps) {
+      super(props);
+    }
+
+    render() {
+      let classNames: string[] = [];
+
+      // Keep existing className attribute if given (either string or array)
+      if (typeof this.props.className === 'string') {
+        classNames.push(this.props.className);
+      } else if (Array.isArray(this.props.className)) {
+        classNames.push(this.props.className.join(' '));
+      }
+
+      // wholeName includes parent, e.g. `parent-block__child-element`
+      const wholeName = parent ? `${parent.blockName}__${name}` : name
+      classNames.push(wholeName);
+
+      // modifiers could be a single string or array of strings or object with booleans
+      if (typeof this.props.m === 'string' && this.props.m.length >= 1) {
+        // Case 1: string
+        classNames.push(`${wholeName}--${this.props.m}`);
+      } else if (Array.isArray(this.props.m)) {
+        // Case 2: array
+        this.props.m.forEach((modifier) => {
+          if (typeof modifier === 'string' && modifier.length >= 1) {
+            classNames.push(`${wholeName}--${modifier}`);
+          }
+        });
+      } else if (typeof this.props.m === 'object') {
```

Comment:
Should we make this a custom type? This reads like we're expecting each entry to have two elements, and if this was something like `{'fb'}` would this result in class names `wholeName__f` and  `wholeName__b`?

Reply:
So the `m` property has these types:

```ts
type BemModifiersObject = {
  [modifierName: string]: boolean
}

interface BemComponentProps extends React.ComponentProps<any> {
  m?: string | string[] | BemModifiersObject
}
```

Reply:
And passing `null` crashes it because it is not a valid value for `m`. Unfortunatly it's not showing up because `editableForm` is not TypeScript yet.

Reply:
If I make this simple code:

```ts
import React from 'react';
import bem, {makeBem} from 'js/bem';

bem.FooBar = makeBem(null, 'foo-bar');

function foo() {
  const isAsideVisible = true;

  return (
    <bem.FooBar m={isAsideVisible ? 'visible' : null} />
  )
}
```

Reply:
I get TypeScript error: `Type 'string | null' is not assignable to type 'string | string[] | BemModifiersObject | undefined'. Type 'null' is not assignable to type 'string | string[] | BemModifiersObject | undefined'.`

Reply:
There is a useful thing in TypeScript - it can understand JSDoc comments and even display them. So I've improved the definition of properties to this:

```ts
interface BemComponentProps extends React.ComponentProps<any> {
  /**
   * Pass a string, or array of strings. For complex modifiers, pass an object:
   * {
   *   'modifier-name': <boolean value whether the modifier should be applied>,
   *   'another-modifier': <boolean>,
   * }
   */
  m?: string | string[] | BemModifiersObject
}
```

And now whenever I hover over `m` in my editor, I see this:

<img width="755" alt="Screenshot 2021-09-14 at 16 51 18" src="https://user-images.githubusercontent.com/2521888/133291666-fbf6a6cd-4a4f-41da-a33b-e9cc6f6cb0c4.png">